File: frontend/src/api/subdivisions.ts
import client from "./client";
import type { SubDivision, SubDivisionCreate, SubDivisionList, SubDivisionMember, SubDivisionUpdate } from "../types";

export async function listSubdivisions(page = 1, size = 10, search?: string): Promise<SubDivisionList> {
  const params: Record<string, string | number> = { page, size };
  if (search) params.search = search;
  const { data } = await client.get<SubDivisionList>("/v1/subdivisions", { params });
  return data;
}

export async function createSubdivision(payload: SubDivisionCreate): Promise<SubDivision> {
  const { data } = await client.post<SubDivision>("/v1/subdivisions", payload);
  return data;
}

export async function getSubdivision(id: number): Promise<SubDivision> {
  const { data } = await client.get<SubDivision>(`/v1/subdivisions/${id}`);
  return data;
}

export async function updateSubdivision(id: number, payload: SubDivisionUpdate): Promise<SubDivision> {
  const { data } = await client.put<SubDivision>(`/v1/subdivisions/${id}`, payload);
  return data;
}

export async function deleteSubdivision(id: number): Promise<void> {
  await client.delete(`/v1/subdivisions/${id}`);
}

export async function addMember(subdivisionId: number, payload: { employee_id: number }): Promise<SubDivisionMember> {
  const { data } = await client.post<SubDivisionMember>(`/v1/subdivisions/${subdivisionId}/members`, payload);
  return data;
}

export async function removeMember(subdivisionId: number, memberId: number): Promise<void> {
  await client.delete(`/v1/subdivisions/${subdivisionId}/members/${memberId}`);
}

File: backend/app/api/v1/subdivisions.py
from fastapi import Depends, Query, status
from fastapi.routing import APIRouter

from app.auth.permissions import Permission, require_permission
from app.schemas.auth import UserContext
from app.schemas.subdivisions import (
    SubDivisionCreate,
    SubDivisionList,
    SubDivisionMemberAdd,
    SubDivisionMemberResponse,
    SubDivisionResponse,
    SubDivisionUpdate,
)
from app.services.subdivisions import SubDivisionService
from app.svc_dependencies import get_subdivision_service

router = APIRouter()


@router.get("", response_model=SubDivisionList)
async def list_subdivisions(
    page: int = Query(1, ge=1),
    size: int = Query(10, ge=1, le=100),
    search: str | None = Query(None),
    _: UserContext = require_permission(Permission.EMPLOYEES_READ),
    service: SubDivisionService = Depends(get_subdivision_service),
):
    subdivisions, total = await service.list(page=page, size=size, search=search)
    return SubDivisionList(items=subdivisions, total=total, page=page, size=size)


@router.post("", response_model=SubDivisionResponse, status_code=status.HTTP_201_CREATED)
async def create_subdivision(
    data: SubDivisionCreate,
    _: UserContext = require_permission(Permission.EMPLOYEES_WRITE),
    service: SubDivisionService = Depends(get_subdivision_service),
):
    subdivision = await service.create(data=data)
    return await service.get_with_members(subdivision_id=subdivision.id)


@router.get("/{subdivision_id}", response_model=SubDivisionResponse)
async def get_subdivision(
    subdivision_id: int,
    _: UserContext = require_permission(Permission.EMPLOYEES_READ),
    service: SubDivisionService = Depends(get_subdivision_service),
):
    return await service.get_with_members(subdivision_id=subdivision_id)


@router.put("/{subdivision_id}", response_model=SubDivisionResponse)
async def update_subdivision(
    subdivision_id: int,
    data: SubDivisionUpdate,
    _: UserContext = require_permission(Permission.EMPLOYEES_WRITE),
    service: SubDivisionService = Depends(get_subdivision_service),
):
    await service.update(subdivision_id=subdivision_id, data=data)
    return await service.get_with_members(subdivision_id=subdivision_id)


@router.delete("/{subdivision_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_subdivision(
    subdivision_id: int,
    _: UserContext = require_permission(Permission.EMPLOYEES_WRITE),
    service: SubDivisionService = Depends(get_subdivision_service),
):
    await service.delete(id=subdivision_id)


@router.post(
    "/{subdivision_id}/members",
    response_model=SubDivisionMemberResponse,
    status_code=status.HTTP_201_CREATED,
)
async def add_member(
    subdivision_id: int,
    data: SubDivisionMemberAdd,
    _: UserContext = require_permission(Permission.EMPLOYEES_WRITE),
    service: SubDivisionService = Depends(get_subdivision_service),
):
    return await service.add_member(subdivision_id=subdivision_id, data=data)


@router.delete(
    "/{subdivision_id}/members/{member_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def remove_member(
    subdivision_id: int,
    member_id: int,
    _: UserContext = require_permission(Permission.EMPLOYEES_WRITE),
    service: SubDivisionService = Depends(get_subdivision_service),
):
    await service.remove_member(subdivision_id=subdivision_id, member_id=member_id)
